---
title: "Cisplatin tox analysis_2"
output: html_notebook
date: "25/08/21"
author: "Finlay Slorach"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load data 

Load saved R objects e.g. previous SCE objects and seurat objects. Get corrected (sequencing depth adjusted) counts for differential expression analysis.

```{r warning=FALSE, message=FALSE}
################## LOAD DATA ############################
root <- '/hpc/scratch/hdd2/fs541623/scRNAseq/Run1_3_Cisplatin_080721/Secondary_analysis'
s <- readRDS(paste0(root, '/new_SCE_objects/FINAL_WORKFLOW/dim_reduction_obj.rds'))
sce_3 <- readRDS(paste0(root, '/new_SCE_objects/FINAL_WORKFLOW/raw_sce.rds'))
counts_corrected <- GetAssayData(s, assay='SCT', slot='counts')
load_packages()
annotations <- load_annotations()
theme_set(theme_bw())
s
```

## Differential expression analysis

Non-parametric difference in mean test. [sctransform:diff mean test]: (https://rawgit.com/ChristophH/sctransform/supp_html/supplement/np_diff_mean_test.html). Here we do multiple pairwise comparisons between timepoints to find conserved differentially expressed genes for each timepoint. Later we also do differential expression comparing each timepoint to the control. In this experiment we only had one control. If there were undosed controls for each timepoint then i think it would be best to compare each dosed group at a given timpeoint to its equivalent undosed. But doing an 'each vs rest' comparison, like we do in this chunk of code, gives upregulated genes specific to each timepoint. 

**Differential expression (DE)**:

- DE for scRNAseq mostly use DE tests that are also used for bulk RNA sequencing data (MAST, DESeq2)
- The gene counts are either aggregated or averaged within a given group e.g. timepoints for more statistical 'power'
- Then each genes expression is compared to every other genes expression
- Because you doing multiple testing of 1000s of genes to 1000s of other genes you need to adjust the signifcance value of 
   your results as the random chance of gene expression being signficantly different from another genes expression increases.
   (Multiple testing correction)
I havn't had the chance to try out MAST but it seems to perform well in benchmark tests [best practices in scRNAseq]:(https://www.embopress.org/doi/full/10.15252/msb.20188746). The documentation is here [MAST]:(https://www.bioconductor.org/packages/release/bioc/vignettes/MAST/inst/doc/MAITAnalysis.html) and you can also do gene set enrichment using this to look at what the DE genes are involved in. 

```{r, message=FALSE, warning=FALSE}
############### DEG ACROSS TIME ########################
de_res <- sctransform::diff_mean_test(
  y=GetAssayData(s, assay='SCT', slot='counts'), group_labels=s$sample, only_pos = F, mean_th=0.25)
#de_res <- readRDS(paste0(root, '/new_SCE_objects/FINAL_WORKFLOW/deg_cellthr2.rds'))

###### get annotations ######
de_res.anno <- annotations %>% 
  dplyr::filter(gene_name %in% de_res$gene) %>% 
  dplyr::rename(gene='gene_name') %>%
  left_join(., de_res, by='gene') %>% 
  arrange(group1, pval_adj) 
#write.csv(de_res.anno, paste0(root, '/new_tables/FINAL_WORKFLOW/deg_cellsthr2annotated.csv'))

head(de_res.anno)
```

Here we also compare each timepoint to the first undosed control

```{r warning=FALSE, message=FALSE}
#### DEG across controls ####
times <- c('0h','2h','8h','16h','24h','48h','72h')
DEG.between_contr <- 
  lapply(times[2:7], function(x) sctransform::diff_mean_test(
  y=counts_corrected, group_labels=s$sample, only_pos = F, 
  compare=c(x, '0h'), mean_th=0.25))
de_res_contr <- do.call(rbind, DEG.between_contr)
de_res_contr$group1 <- factor(de_res_contr$group1)
#saveRDS(de_res_contr, paste0(root, '/new_SCE_objects/FINAL_WORKFLOW/deg_cellthr2_control.rds'))

###### get annotations ###### 
de_res.anno.c <- annotations %>% 
  dplyr::filter(gene_name %in% de_res_contr$gene) %>% 
  dplyr::rename(gene='gene_name') %>%
  left_join(., de_res_contr, by='gene') %>% 
  arrange(group1, pval_adj) 
#write.csv(de_res.anno, paste0(root, '/new_tables/FINAL_WORKFLOW/deg_cellsthr2annotated_control.csv'))
```

For pathway analysis in Ingenuity pathway analysis (IPA) the DE genes need to be formatted correctly. This chunk gives a commmon set of genes for all the timepoints for use in IPA and also gets th ENSEMBL gene ids

```{r warning=FALSE, messages=FALSE}
### get gene ids ###
de_res.anno.c$gene.ids <- mapIds(org.Hs.eg.db, keys =de_res.anno.c$gene, column='ENSEMBL', keytype='SYMBOL')
de_res.anno.c <- de_res.anno.c %>% dplyr::select(gene, group1, pval_adj,log2FC, gene.ids)
de_res.anno.c <- de_res.anno.c %>% split.data.frame(factor(.$group1)) %>% purrr::reduce(merge, by='gene')
#write.csv(de_res_anno, paste0(root, '/new_tables/FINAL_WORKFLOW/deg_cellsthr2_control_IPAformatted.csv'))

```

Comparing each timepoint to eachother rather than to the control seemed to give better results. Here I write the DE genes to a csv for IPA

```{r warning=FALSE, message=FALSE}
####### use vs time ######
### get gene ids and format for IPA  ###
de_res.anno$gene.ids <- mapIds(org.Hs.eg.db, keys =de_res.anno$gene, column='ENSEMBL', keytype='SYMBOL')
de_res.anno <- de_res.anno %>% dplyr::select(gene, group1, pval_adj,log2FC, gene.ids)
de_path_rest <- de_res %>% split.data.frame(factor(.$group1)) %>% purrr::reduce(merge, by='gene')
#write.csv(de_path_rest, paste0(root, '/new_tables/FINAL_WORKFLOW/deg_cellsthr2_IPA.csv'))
```

Then we can produce two commonly seen plots used in -omics analysis; volcano plots, and heatmaps. 

- Volcano plots show the signifance of each gene from DE analysis and their log2FC. -log10(pval_adjusted) is the y axis so that the     most significant genes have the highest y axis values. In the sctransform vignette linked previously they also use z-score which is   a measure of dispersion from a mean gene expression value. 
- Heatmaps show the scaled gene expression values in a compact form.

```{r fig.width=10, fig.height=7}
### VOLCANO PLOT ###
de_res <- readRDS(paste0(root, '/new_SCE_objects/FINAL_WORKFLOW/deg_cellthr2.rds'))
de_res_filt <- de_res %>% dplyr::filter(group1 %in% c('2h','8h','16h','24h','48h','72h'))
markers_vol <- group_by(de_res, group1) %>% 
  dplyr::filter(rank(pval_adj, ties.method='first') <= 6) %>%  
  dplyr::select(group1, gene, mean1, mean2, log2FC, zscore, emp_pval_adj, pval_adj)  %>% 
  dplyr::filter(group1 %in% c('2h','8h','16h','24h','48h','72h'))
### Plot most significant genes ###
ggplot(de_res_filt, aes(x=pmin(log2FC,10), y=pmin(-log10(pval_adj), 10))) + 
  geom_point(aes(color = pval_adj < 0.05))  + 
  geom_point(data=markers_vol, color='deeppink') +
  geom_text_repel(data = markers_vol, mapping = aes(label = gene), max.overlaps=Inf, force=3) + 
  theme(legend.position = "bottom") + 
  facet_wrap(~group1, ncol = 3) + ylab("-log10(pval_adj)") + 
  geom_vline(xintercept=0, linetype='dashed', colour='#777777', size=0.2) +
  geom_hline(yintercept=0, linetype='dashed', colour='#777777', size=0.2) +
  xlab("log2 mean fold-change")
#ggsave(paste0(root, '/new_visualisations/FINAL_WORKFLOW/volcano_top6_log10pvaladj.png'), height=10, width=10)
```

Here we get the most signficant genes and plot them on a seurat heatmap, for a quick check, and only show the treated groups.

```{r warning=FALSE, messages=FALSE}
############ HEATMAP ############################
#marker_genes <- group_by(de_res, group1) %>% dplyr::filter(rank(pval_adj, ties.method = "first") <= 15)  
#s <- GetResidual(s, features=marker_genes$gene, verbose=F)
sig.genes <- de_res %>% dplyr::filter(pval_adj < 0.05) %>% dplyr::filter(group1!='0h')
## plot most upregulated genes ##
s.s <- subset(s, subset=sample!='0h')
s.s <- GetResidual(s.s, features=sig.genes$gene, verbose=F)
DoHeatmap(s.s, group.by='sample', features=sig.genes$gene, slot = "scale.data") + 
scale_fill_gradient2(low = "blue", mid = "white", high = "#FF2800")
```

To have more control over the layout of the heatmap we can plot a complex heatmap [complex heatmap applied to scRNAseq]:(https://jokergoo.github.io/ComplexHeatmap-reference/book/more-examples.html#visualize-cell-heterogeneity-from-single-cell-rnaseq). This allows us to pick out some interesting genes to label the heatmap. To do this looked through the DE csv file and created an extra label column to label genes involved in DNA damage or platinum resistance etc and then retrieved these for plotting.

```{r results='hide'}
## complex heatmap ##
# // commented code involving pseudotime will only work if trajectory analysis R script has been ran // # 
#s$pseudotime <- sce$slingPseudotime_1
#annotations <- FetchData(s, vars=c('sample', 'pseudotime', 'mito.percent'))
annotations <- FetchData(s, vars=c('sample', 'mito.percent'))
#colnames(annotations) <- c('Time', 'psuedotime', 'mitochondrial.percent')
colnames(annotations) <- c('Time', 'mitochondrial.percent')
ha.anno <- HeatmapAnnotation(df=annotations, 
    col = list(Time=c('0h'=cc[1], '2h'=cc[2],'8h'=cc[3],'16h'=cc[4],'24h'=cc[5],'48h'=cc[6],'72h'=cc[7]),
    # psuedotime=colorRamp2(c(0,0.75, 1.5), c("green", "white", "blue")),
    mitochondrial.percent=colorRamp2(c(0, 15, 30), c("white", "red", "darkred"))
    ))
markers <- read.csv(paste0(root, '/new_tables/FINAL_WORKFLOW/Copy of Copy of deg_cellsthr2annotated_170821.csv'))
plat.genes <- markers %>% dplyr::filter(LABEL == 'Platinum') %>% pull(symbol)
dna.gene <- markers %>% dplyr::filter(LABEL == 'DNA') %>% pull(symbol)
markers <- markers[!(markers %in% c(dna.gene, plat.genes))][1:20]
mark.toplot <- markers[markers$LABEL == '*',] %>% pull(symbol)
counts <- GetAssayData(s, assay='SCT', slot='scale.data')
counts.mat <- counts[rownames(counts) %in% sig.genes$gene,]
markers <- c(plat.genes, dna.gene, unique(mark.toplot))
#plot.new()
#png(paste0(root, '/new_visualisations/FINAL_WORKFLOW/DE_complex_heatmap1.png'), height=680, width=480)
Heatmap(counts.mat,  
    column_split = factor(s$sample),
    column_title = 'Differentially expressed genes across time',
    name='Expression',
    col = colorRamp2(c(-2, 0, 2), c("blue", "white", "#FF2800")),
    cluster_columns = F,
    show_column_dend = FALSE,
    cluster_column_slices = TRUE,
    column_title_gp = gpar(fontsize = 12),
    column_gap = unit(0.5, "mm"),
    width=unit(8,'cm'),
    height=unit(15, 'cm'),
    cluster_rows = F,
    show_row_dend = T,
    column_title_rot = 0,
    top_annotation = ha.anno,
    show_column_names = FALSE,
    show_row_names = F) +
rowAnnotation(link=anno_mark(at = which(rownames(counts.mat) %in% markers),
    labels = (rownames(counts.mat)[rownames(counts.mat) %in% markers]), 
    labels_gp = gpar(fontsize = 9, col='black'), padding = unit(0.5, "mm")))
```

```{r warning=FALSE}
knitr::include_graphics(paste0(root, '/new_visualisations/FINAL_WORKFLOW/DE_complex_heatmap1.png'), )
```



```{r}
##### SAVE OBJECT #####
#saveRDS(s, paste0(root, '/new_SCE_objects/FINAL_WORKFLOW/DE_obj.rds'))
```

